ref/go/src/bech32/bech32.go

```diff
@@ -0,0 +1,197 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package bech32
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+)
+
+// data part characters
+var CHARSET = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
+
+var generator = []int{0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3}
+
+func polymod(values []int) int {
+	chk := 1
+	for _, v := range values {
+		top := chk >> 25
+		chk = (chk&0x1ffffff)<<5 ^ v
+		for i := 0; i < 5; i++ {
+			if (top>>uint(i))&1 == 1 {
+				chk ^= generator[i]
+			}
+		}
+	}
+	return chk
+}
+
+func hrpExpand(hrp string) []int {
+	ret := []int{}
+	for _, c := range hrp {
+		ret = append(ret, int(c>>5))
+	}
+	ret = append(ret, 0)
+	for _, c := range hrp {
+		ret = append(ret, int(c&31))
+	}
+	return ret
+}
+
+func verifyChecksum(hrp string, data []int) bool {
+	return polymod(append(hrpExpand(hrp), data...)) == 1
+}
+
+func createChecksum(hrp string, data []int) []int {
+	values := append(append(hrpExpand(hrp), data...), []int{0, 0, 0, 0, 0, 0}...)
+	mod := polymod(values) ^ 1
+	ret := make([]int, 6)
+	for p := 0; p < len(ret); p++ {
+		ret[p] = (mod >> uint(5*(5-p))) & 31
+	}
+	return ret
+}
+
+// Encode encodes hrp(human-readable part) and data(32bit data array), returns Bech32 / or error
+func Encode(hrp string, data []int) (string, error) {
+	combined := append(data, createChecksum(hrp, data)...)
+	var ret bytes.Buffer
+	ret.WriteString(hrp)
+	ret.WriteString("1")
+	for idx, p := range combined {
+		if p < 0 || p >= len(CHARSET) {
+			return "", fmt.Errorf("invalid data : data[%d]=%d", idx, p)
+		}
+		ret.Write([]byte{CHARSET[p]})
+	}
+	_, _, err := Decode(ret.String())
+	if err != nil {
+		return "", err
+	}
+	return ret.String(), nil
+}
+
+// Decode decodes bechString(Bech32) returns hrp(human-readable part) and data(32bit data array) / or error
+func Decode(bechString string) (string, []int, error) {
+	if len(bechString) > 90 {
+		return "", nil, fmt.Errorf("too long : len=%d", len(bechString))
+	}
+	if strings.ToLower(bechString) != bechString && strings.ToUpper(bechString) != bechString {
+		return "", nil, fmt.Errorf("mixed case")
+	}
+	bechString = strings.ToLower(bechString)
+	pos := strings.LastIndex(bechString, "1")
+	if pos < 1 || pos+7 > len(bechString) {
+		return "", nil, fmt.Errorf("separator '1' at invalid position : pos=%d , len=%d", pos, len(bechString))
+	}
+	for p := 0; p < pos; p++ {
```

Comment:
If you move `hrp := bechString[0:pos]` here, you do `for i, char := range hrp {` to avoid manual indexing.